test/libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,1989 @@
+			{"Entry", "function E.callSuperH()"},
+			{"function E.callF()", "function C.f()"},
+			{"function E.callG()", "function E.g()"},
+			{"function E.callH()", "function E.h()"},
+			{"function E.callI()", "function E.i()"},
+			{"function E.callCF()", "function C.f()"},
+			{"function E.callCG()", "function C.g()"},
+			{"function E.callCH()", "function C.h()"},
+			{"function E.callDH()", "function D.h()"},
+			{"function E.callEI()", "function E.i()"},
+			{"function E.callSuperF()", "function C.f()"},
+			{"function E.callSuperG()", "function C.g()"},
+			{"function E.callSuperH()", "function D.h()"},
+		}},
+	};
+
+	checkCallGraphExpectations(get<0>(graphs), expectedCreationEdges);
+	checkCallGraphExpectations(get<1>(graphs), expectedDeploymentEdges);
+}
+
+BOOST_AUTO_TEST_CASE(overloaded_functions)
+{
+	unique_ptr<CompilerStack> compilerStack = parseAndAnalyzeContracts(R"(
+		enum E {E1, E2, E3}
+
+		function free() {}
+		function free(uint) {}
+		function free(bytes memory) {}
+		function free(E) {}
+
+		contract C {
+			function f(E) internal {}
+			function f(bool) external {}
+		}
+
+		contract D is C {
+			function ext1() external { free(); free(123); free("123"); }
+			function ext2() external { f(); f(123); f("123"); }
+			function ext3() external { free(E.E2); f(E.E2); }
+			function ext4() external { this.f(false); }
+
+			function f() internal {}
+			function f(uint) internal {}
+			function f(bytes memory) internal {}
+		}
+	)"s);
+	tuple<CallGraphMap, CallGraphMap> graphs = collectGraphs(*compilerStack);
+
+	map<string, EdgeNames> expectedCreationEdges = {
+		{"C", {}},
+		{"D", {}},
+	};
+
+	map<string, EdgeNames> expectedDeploymentEdges = {
+		{"C", {
+			{"Entry", "function C.f(bool)"},
+		}},
+		{"D", {
+			{"Entry", "function C.f(bool)"},
+			{"Entry", "function D.ext1()"},
+			{"Entry", "function D.ext2()"},
+			{"Entry", "function D.ext3()"},
+			{"Entry", "function D.ext4()"},
+			{"function D.ext1()", "function free()"},
+			{"function D.ext1()", "function free(uint256)"},
+			{"function D.ext1()", "function free(bytes)"},
+			{"function D.ext2()", "function D.f()"},
+			{"function D.ext2()", "function D.f(uint256)"},
+			{"function D.ext2()", "function D.f(bytes)"},
+			{"function D.ext3()", "function free(enum E)"},
+			{"function D.ext3()", "function C.f(enum E)"},
+		}},
+	};
+
+	checkCallGraphExpectations(get<0>(graphs), expectedCreationEdges);
+	checkCallGraphExpectations(get<1>(graphs), expectedDeploymentEdges);
+}
+
+BOOST_AUTO_TEST_CASE(modifiers)
+{
+	unique_ptr<CompilerStack> compilerStack = parseAndAnalyzeContracts(R"(
+		library L {
+			modifier m() { g(); _; }
+
+			function f() m internal {}
+			function g() internal {}
+		}
+
+		contract C {
+			modifier m1() virtual { _; }
+
+			function q() m1 internal virtual { L.f(); }
+		}
+
+		contract D is C {
+			modifier m2() { q(); _; new C(); }
+
+			function p() m2 internal { C.q(); }
+			function q() m2 internal override virtual {}
+		}
```

Comment:
Note: currently a modifier is included in the graph as something called by the function it modifies. For example in the above we have `p() -> m2`, `p() -> C.q()`, `m2 -> q()` and `C.q() -> m1` (ignoring other edges).

But the actual execution order is different. It's more like `m2 -> p() -> m1 -> C.q()` because the modifier wraps function body. Should I change that?

Reply:
I talked about it with @chriseth.

My suggestion of putting the modifier in between the call and the function has the downside that it would look like the modifier is calling all the functions that use it which would be misleading (there are separate inlined instances of the modifier that do not have anything to do with each other). If we wanted to do this, it would be better to just "inline" the modifier in the graph too - making it look like the function is doing all the stuff the modifier does.

@chriseth suggested that the way it's done currently is probably fine though. After giving it some thought I think I agree. It does have an effect similar to "inlining" the modifier in the graph without actually hiding it.